### PR TITLE
Add post-processing scripts for the ISMIP6 experiment

### PR DESCRIPTION
**Pull request by hollyhan**
This PR adds scripts that post-processes MALI state and flux output variables in the format required by the ISMIP6 team. In the previous (first) participation of the MALI team to the ISMIP6-2100 experimental protocol, post-processing MALI outputs involved many steps of manual running of several python scripts and `nco` commands, which can be tedious and exposed to potential human errors. The scripts here have been written to improve the automation of the workflow and thus reduce the room for making mistakes.  

This set of scripts automate the process.  There are separate steps for processing 2d state fields, 2d flux fields, and scalar fields.  Within each step, there are operations for adjusting the raw MALI output if needed, time-averaging or time-subsetting as appropriate, and remapping to the ISMIP6 grid (as appropriate).  It is recommended that post-processing be done on a compute/interactive node, as some of the steps can be time consuming.

### CODE REVIEW

**Comment by hollyhan** [comment]
Hi @matthewhoffman, the PR has been opened now! 
If it helps (since there are many flags one needs to provide), here is the sample command line I used to run the scripts:
```
./post_process_mali_to_ismip6.py -i_state output_state_2015.nc -i_flux output_flux_all.nc -i_init AIS_8to30km_r01_20220906.smooth3.basinsFineTuned.nc --mali_mesh_name AIS_8to30km --ismip6_grid_file UKESM1-0-LL_temperature_8km_x_60m.nc  -e ctrl -g globalStats.nc  -p /Users/hollyhan/ISMIP6-PostProcessing/files_states_processed/ 
```
Note that you might run into errors running the script if the MALI initial and output (state and flux) files that you use to test this PR are missing the variables required by the ISMIP6. (There are some lines that I commented out because the output files I was using did not have all of the variables required by the ISMIP6.).

File: landice/output_processing_li/process_state_variables.py [hunk 1]
```diff
@@ -0,0 +1,618 @@
...
+        vol_snapshot[i] = vol_yearly[0, i]
+        vaf_snapshot[i] = vaf_yearly[0, i]
+        gia_snapshot[i] = gia_yearly[0, i]
+        fia_snapshot[i] = fia_yearly[0, i]
+        smbi = smb_yearly[:, i]
+        bmbi = bmb_yearly[:, i]
+        cfxi = cfx_yearly[:, i]
+        gfxi = gfx_yearly[:, i]
+
+        # take the average of the flux variables
```
Comment (matthewhoffman) [inline landice/output_processing_li/process_state_variables.py:399]: These should ideally be a weighted average: `(var[:]*deltat[:]).sum()/deltat[:].sum()` for the indices of each year.

File: landice/output_processing_li/process_flux_variables.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+
+    # create a temporary file to which processed data will be saved with the
+    # right dimension
+    ind = len(years) - 1
+    command = ["ncks", "-O",
+               "-d", f"Time,0,{str(ind)}",
+               input_file,
+               "tmp.nc"]
+
+    check_call(command)
```
Comment (matthewhoffman) [inline landice/output_processing_li/process_flux_variables.py:55]: This time we are writing one flux file per year, so I don't think we need to do any of this.  We simply need to open each file, perform the weighted average over all time levels in that file, and save the result.

**Review by matthewhoffman — CHANGES_REQUESTED** [review]
@hollyhan , thank you so much for getting these processing scripts organized and cleaned up.  This will be a huge help going forward.  In my comments, I've tried to balance having clean, reusable code with the fact that it's possible we won't be running this frequently or any time soon (only when ISMIP activities occur).  But I'm assuming there may be future ISMIP activities in the coming years that we'll want to participate in, so we should aim for this to be flexible beyond the current set of experiments, where possible without too much trouble.  

Most of comments are pretty minor, but one fairly big one I see is that I think it will be important to be able to point the scripts to a directory without having to ncks all the individual output files together first.  However, that change is fairly fundamental, so I'll follow up with you about if/how we can handle that.

File: landice/output_processing_li/process_flux_variables.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+    areaCell = data['areaCell'][:].values
+    deltat = data['deltat'][:].values
+    # edgeMask = data['edgeMask'][:, :].values # this needs to be outputted as well. Once commented out, comment out L195 as well, and indent L197
+    # dHdt = data['dHdt'][:, :].values # Uncomment this line once the dHdt variable is outputted in the stream and delete the line below
+    dHdt = data['calvingThickness'][:, :].values  # delete this line once the dHdt variable is outputted in the stream. This line is used just for now to check the code
+    fluxGLEdge = data['fluxAcrossGroundingLine'][:, :]
+    fluxGLCell_array = np.zeros((time, nCells))
+
+    print("=== starting the grounding line flux processing ===")
+    nCells=10
```
Comment (hollyhan) [inline landice/output_processing_li/process_flux_variables.py:132]: remove

File: landice/output_processing_li/process_flux_variables.py [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+                calvingVelDivideThickness = 0.0
+                # unit: m^2/s
+
+            calvingFlux = rho_i * calvingVelDivideThickness
+            # unit: kg/m/s
+            calvingFluxArray[t, i] = calvingFlux
+
+    data['calvingFlux'] = calvingFluxArray
+    data.to_netcdf(
+        file_flux_on_cell)  # writing out to a netCDF file seems to be needed to save the newly added variable `fluxAcrossGroundingLineCell`.
```
Comment (hollyhan) [inline landice/output_processing_li/process_flux_variables.py:198]: clean up

File: landice/output_processing_li/post_process_mali_to_ismip6.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    parser = argparse.ArgumentParser(
+                        description='process MALI outputs for the ISMIP6'
+                                    'submission')
+    parser.add_argument("-e", "--exp_name", dest="exp",
+                        required=True,
+                        help="ISMIP6 experiment name (e.g., exp05")
+    parser.add_argument("-i_state", "--input_state", dest="input_file_state",
+                        required=True, help="mpas output state variables")
+    parser.add_argument("-i_flux", "--input_flux", dest="input_file_flux",
+                        required=False, help="mpas output flux variables")
```
[the remark this answers is not in the document]
Comment (matthewhoffman) [reply landice/output_processing_li/post_process_mali_to_ismip6.py:32]: After our discussion, we'll defer this to a future PR if needed.

File: landice/output_processing_li/process_flux_variables.py (hunk 4 above)
Comment (matthewhoffman) [reply landice/output_processing_li/process_flux_variables.py:198]: I'm commenting the calving chunk for now.  We can remove if we add the calving calculation directly to MALI.  Or clean it up if we don't.

File: landice/output_processing_li/process_state_variables.py [hunk 6]
```diff
@@ -0,0 +1,618 @@
...
+    yValues = dataOut.createVariable('y', 'f4', ('y'))
+    timeValues = dataOut.createVariable('time', 'f4', ('time'))
+    AUTHOR_STR = 'Matthew Hoffman, Trevor Hillebrand, Holly Kyeore Han'
+    DATE_STR = '23-Oct-2022'
+
+    for i in range(timeSteps):
+
+        timeValues[i] = i * timeSpan * spy
+
+        if not ismip6_var_name == 'sftgif':
```
[the remark this answers is not in the document]
Comment (matthewhoffman) [reply landice/output_processing_li/process_state_variables.py]: ISMIP6 protocol states: "All "missing data" must be assigned the single-precision floating point value of 1.e20. Fields should be undefined outside the ice mask."  So I'm leaving as is.

File: landice/output_processing_li/ismip6_postprocessing/process_flux_variables.py [hunk 7]
```diff
@@ -140,106 +143,80 @@ def do_time_avg_flux_vars(input_file, output_file):
...
+
+    # get calving thickness values
+    calvingThickness = data['calvingThickness'][:, :].values
+    assert time == len(deltat)
+
+    # create and initialize a new data array for calvingFluxArray
+    calvingFluxArray = data['calvingThickness'].copy()
+    calvingFluxArray = calvingFluxArray * 0.0
+
+    # This part seems to have been used to get `edgeMask`
```
Comment (trhille) [inline landice/output_processing_li/ismip6_postprocessing/process_flux_variables.py:180]: Can we delete this commented section?

File: landice/output_processing_li/ismip6_postprocessing/process_flux_variables.py [hunk 8]
```diff
@@ -140,106 +143,80 @@ def do_time_avg_flux_vars(input_file, output_file):
...
+        index_cf = np.where(calvingThickness[t, :] > 0)[0]
+        for i in index_cf:
+
+            ne = nEdgesOnCell[i]
+            cliffArea = 0.0
+            for j in range(ne):
+                neighborCellId = cellsOnCell[i, j] - 1
+                neighborEdgeId = edgesOnCell[i, j] - 1
+
+                # if ((edgeMask[0, edgeId[j]] & 16) / 16 and (edgeMask[0, edgeId[j]] & 4) / 4):  # uncomment this line and indent L197 once edgeMask is acquired.
```
Comment (trhille) [inline landice/output_processing_li/ismip6_postprocessing/process_flux_variables.py:200]: can we delete?

**Review by trhille — COMMENTED** [review]
@matthewhoffman, the changes in L188–219 in `process_flux_variables.py` make sense to me for the `restore_calving_front` case. I marked a few places where I think commented legacy code should be removed for clarity.

File: landice/output_processing_li/ismip6_postprocessing/process_flux_variables.py (hunk 7 above)
Comment (matthewhoffman) [reply landice/output_processing_li/ismip6_postprocessing/process_flux_variables.py]: Yes, done in ae2ad7d

File: landice/output_processing_li/ismip6_postprocessing/process_flux_variables.py (hunk 8 above)
Comment (matthewhoffman) [reply landice/output_processing_li/ismip6_postprocessing/process_flux_variables.py]: yes, done in ae2ad7d

File: landice/output_processing_li/ismip6_postprocessing/process_flux_variables.py [hunk 9]
```diff
@@ -0,0 +1,446 @@
...
+                thresholdBoundaryContributors[ownerIdx] += 1
+            #print(thresholdBoundaryAssignedVolume.sum(), (calvingThicknessFromThreshold[t,:]*areaCell[:]).sum())
+            assert np.absolute(thresholdBoundaryAssignedVolume.sum() - (calvingThicknessFromThreshold[t,:]*areaCell[:]).sum()) < 1.0
+            #for i in bdyIndices:
+                #print(f"length={thresholdBoundaryLength[i]}, vol={thresholdBoundaryAssignedVolume[i]}, sumthk={thresholdBoundarySummedThickness[i]}, num={thresholdBoundaryContributors[i]}, meanthk={thresholdBoundarySummedThickness[i]/thresholdBoundaryContributors[i]}")
+            # Finally calculate licalvf for each boundary cell and add to whatever was already there
+            thresholdSpeed = thresholdBoundaryAssignedVolume / \
+                    (thresholdBoundarySummedThickness/thresholdBoundaryContributors * \
+                     thresholdBoundaryLength) / \
+                    deltat[t]    # units of kg / m2 / s
```
Comment (trhille) [inline landice/output_processing_li/ismip6_postprocessing/process_flux_variables.py:264]: The units here are not yet kg /m2/s, but m/s

File: landice/output_processing_li/ismip6_postprocessing/process_flux_variables.py [hunk 10]
```diff
@@ -0,0 +1,446 @@
...
+            # Our estimated threshold speed is really a retreat speed. So to get calving speed, add on the advective speed
+            thresholdFlux[t, bdyIndices] += (thresholdSpeed[bdyIndices] + surfaceSpeed[t,bdyIndices]) * rho_i
+            calvingFluxArray[t,:] += thresholdFlux[t,:]
```
Comment (trhille) [inline landice/output_processing_li/ismip6_postprocessing/process_flux_variables.py:267]:
```suggestion
            calvingFluxArray[t,bdyIndices] += thresholdFlux[t,bdyIndices]
```
This should amount to the same thing, but I think it's safer than using : indexing.

**Review by trhille — COMMENTED** [review]
@matthewhoffman, I've gone through `clean_flux_fields_before_time_averaging()` and it makes sense to me. I made a few minor comments for clarity.

File: landice/output_processing_li/ismip6_postprocessing/process_flux_variables.py (hunk 10 above)
Comment (matthewhoffman) [reply landice/output_processing_li/ismip6_postprocessing/process_flux_variables.py]: Changed

File: landice/output_processing_li/ismip6_postprocessing/process_flux_variables.py [hunk 11]
```diff
@@ -0,0 +1,446 @@
...
+    # We've encountered a few enormous BMB values.  Until we solve where that is coming from,
+    # set them to something reasonable.
+    floatingBasalMassBalApplied = data['floatingBasalMassBalApplied'][:, :].values
+    for t in range(1, time):
+        if t%20 == 0:
+            print(f"    Time: {t+1} / {time}")
+        # Set large negative BMB values (1 m/s) to the equivalent of the thickness from the previous time step
+        # (Commented line here picked up too many places)
+        #ind = np.nonzero((-floatingBasalMassBalApplied[t,:]/rho_i*deltat[t] > thickness[t-1,:]) * (thickness[t-1,:]>0.0))[0]
+        ind = np.nonzero(floatingBasalMassBalApplied[t,:]/rho_i < -1.0)[0]
```
[the remark this answers is not in the document]
Comment (matthewhoffman) [reply landice/output_processing_li/ismip6_postprocessing/process_flux_variables.py:186]: Yes, the 0 index is to get the first (and only) tuple of actual values.  I double checked that it returns a list longer than 1 when it should.

**Review by matthewhoffman — APPROVED** [review]

**Pull request merged**